front/lib/api/assistant/agent_usage.ts

```diff
@@ -285,38 +164,146 @@ export async function getAgentUsage(
+  workspaceId: number
+): Promise<mentionCount[]> {
+  const mentions = await Mention.findAll({
+    attributes: [
+      "agentConfigurationId",
+      [Sequelize.fn("COUNT", Sequelize.col("mention.id")), "count"],
+    ],
+    where: {
+      createdAt: {
+        [Op.gt]: literal("NOW() - INTERVAL '30 days'"),
+      },
+    },
+    include: [
+      {
+        model: Message,
+        required: true,
+        attributes: [],
+        include: [
+          {
+            model: UserMessage,
+            as: "userMessage",
+            attributes: [],
+            required: true,
+          },
+          {
+            model: Conversation,
+            as: "conversation",
+            attributes: [],
+            required: true,
+            where: {
+              workspaceId: workspaceId,
+            },
+          },
+        ],
+      },
+    ],
+    order: [["count", "DESC"]],
+    group: ["mention.agentConfigurationId"],
+    raw: true,
+  });
+  return mentions.map((mention) => ({
+    agentId: mention.agentConfigurationId as string,
+    count: (mention as unknown as { count: number }).count,
+    timePeriodSec: rankingTimeframeSec,
+  }));
+}
+
+export async function storeCountsInRedis(
+  workspaceId: string,
+  agentMessageCounts: mentionCount[],
+  userCounts: mentionCount[],
+  redis: Awaited<ReturnType<typeof redisClient>>
+) {
+  const transaction = redis.multi();
+  const { agentMessageCountKey, agentUserCountKey } =
+    _getUsageKeys(workspaceId);
+
+  agentMessageCounts.forEach(({ agentId, count }) => {
+    transaction.hSet(agentMessageCountKey, agentId, count);
+  });
+  transaction.expire(
+    agentMessageCountKey,
+    popularityComputationTimeframeSec * 2
+  );
+
+  userCounts.forEach(({ agentId, count }) => {
+    transaction.hSet(agentUserCountKey, agentId, count);
+  });
+  transaction.expire(agentUserCountKey, popularityComputationTimeframeSec * 2);
+
+  const results = await transaction.exec();
+  if (results.includes(null)) {
+    throw new Error("Transaction failed and was rolled back.");
+  }
+}
+
 export async function signalAgentUsage({
   agentConfigurationId,
   workspaceId,
-  userId,
-  timestamp,
-  messageModelId,
 }: {
   agentConfigurationId: string;
   workspaceId: string;
-  userId: string;
-  timestamp: number;
-  messageModelId: ModelId;
 }) {
-  let redis: Awaited<ReturnType<typeof redisClient>> | null = null;
-  try {
-    redis = await redisClient();
-    await populateUsageIfNeeded({
-      agentConfigurationId,
-      workspaceId,
-      messageModelId,
-      redis,
-    });
-    await signalInRedis({
-      agentConfigurationId,
-      workspaceId,
-      userId,
-      timestamp,
-      redis,
-    });
-  } finally {
-    if (redis) {
-      await redis.quit();
-    }
-  }
+  const redis = await redisClient();
+  const { agentMessageCountKey } = _getUsageKeys(workspaceId);
+  await redis.hIncrBy(agentMessageCountKey, agentConfigurationId, 1);
+  await redis.quit();
```

Comment:
If anything above throws you leave a redis open. We need a try catch here + finally